File: app/src/main/java/me/arianb/usb_hid_client/report_senders/KeySender.java
package me.arianb.usb_hid_client.report_senders;

import static me.arianb.usb_hid_client.hid_utils.CharacterDevice.KEYBOARD_DEVICE_PATH;

import android.view.View;

public class KeySender extends ReportSender {
    // Report IDs
    private static final byte STANDARD_KEY = 0x01;
    private static final byte MEDIA_KEY = 0x02;

    public KeySender(View parentLayout) {
        super(KEYBOARD_DEVICE_PATH, true, parentLayout);
    }

    public void addStandardKey(byte modifier, byte key) {
        super.addReportWithLock(new byte[]{STANDARD_KEY, modifier, 0, key, 0});
    }

    public void addMediaKey(byte key) {
        super.addReportWithLock(new byte[]{MEDIA_KEY, key, 0});
    }
}

File: app/src/main/java/me/arianb/usb_hid_client/report_senders/KeySender.kt
package me.arianb.usb_hid_client.report_senders

import me.arianb.usb_hid_client.hid_utils.KeyboardDevicePath

class KeySender(
    keyboardDevicePath: KeyboardDevicePath
) : ReportSender(
    keyboardDevicePath
) {
    fun addStandardKey(modifier: Byte, key: Byte) {
        super.addReportToChannel(byteArrayOf(STANDARD_KEY, modifier, 0, key, 0))
    }

    fun addMediaKey(key: Byte) {
        super.addReportToChannel(byteArrayOf(MEDIA_KEY, key, 0))
    }

    // Every time we send a report, we only send the "key-down" event. This method will automatically send the "key-up"
    // event right afterward
    override fun sendReport(report: ByteArray) {
        // Send "key-down" report
        writeBytes(report)

        // Send "key-up" report of all zeroes (preserving report ID) to release
        val releaseReport = ByteArray(report.size)
        releaseReport[0] = report[0]
        writeBytes(releaseReport)
    }

    companion object {
        // Report IDs
        private const val STANDARD_KEY: Byte = 0x01
        private const val MEDIA_KEY: Byte = 0x02
    }
}
